kessler/kessler_update.F90

```diff
@@ -0,0 +1,118 @@
+module kessler_update
+
+   use ccpp_kinds, only: kind_phys
+
+   implicit none
+   private
+
+   public :: kessler_update_init
+   public :: kessler_update_timestep_init
+   public :: kessler_update_run
+   public :: kessler_update_timestep_final
+
+   ! Private module variables
+   real(kind_phys)    :: gravit
+   real(kind_phys)    :: cpair
+
+CONTAINS
+
+   !> \section arg_table_kessler_update_init  Argument Table
+   !! \htmlinclude kessler_update_init.html
+   subroutine kessler_update_init(gravit_in, cpair_in)
+      real(kind_phys),    intent(in)    :: gravit_in
+      real(kind_phys),    intent(in)    :: cpair_in
+
+      gravit = gravit_in
+      cpair = cpair_in
+
+   end subroutine kessler_update_init
+
+   !> \section arg_table_kessler_update_timestep_init  Argument Table
+   !! \htmlinclude kessler_update_timestep_init.html
+   subroutine kessler_update_timestep_init(temp, temp_prev, ttend_t,          &
+        errmsg, errflg)
+
+      real(kind_phys),    intent(in)  :: temp(:,:)
+      real(kind_phys),    intent(out) :: temp_prev(:,:)
+      real(kind_phys),    intent(out) :: ttend_t(:,:)
+      character(len=512), intent(out) :: errmsg
+      integer,            intent(out) :: errflg
+
+      !   Initialize the previous temperature and its tendency to zero
+      temp_prev(:,:)  = temp(:,:)
+      ttend_t(:,:)    = 0._kind_phys
+
+      errmsg = ''
+      errflg = 0
+
+   end subroutine kessler_update_timestep_init
+
+   !> \section arg_table_kessler_update_run  Argument Table
+   !! \htmlinclude kessler_update_run.html
+   subroutine kessler_update_run(nz, ncol, temp, theta, exner, dt,      &
+        temp_prev, ttend_t, errmsg, errflg)
+
+      integer,            intent(in)    :: nz
+      integer,            intent(in)    :: ncol
+      real(kind_phys),    intent(in)    :: temp(:,:) ! temperature
+      real(kind_phys),    intent(in)    :: exner(:,:)
+      real(kind_phys),    intent(in)    :: dt
+
+      real(kind_phys),    intent(inout) :: theta(:,:)
+      real(kind_phys),    intent(inout) :: temp_prev(:,:)
+      real(kind_phys),    intent(inout) :: ttend_t(:,:)
+
+      character(len=512), intent(out)   :: errmsg
+      integer,            intent(out)   :: errflg
+
+      integer                           :: klev
+      real(kind_phys)                   :: ptend_s(ncol)
+
+      errmsg = ''
+      errflg = 0
+
+      ! Back out tendencies from updated fields
+      do klev = 1, nz
+         ptend_s(:ncol) = ((theta(:ncol,klev) * exner(:ncol,klev)) -          &
```

Comment:
I realize that we don't really want to change the physics calculations themselves during the switch to the CCPP framework, but if `ttend_t` is the only output, then one could calculate that directly without `ptend_s` like so:

`ttend_t(:ncol,klev) = ttend_t(:ncol,klev) + ((theta(:ncol,klev) * exner(:ncol,klev)) - temp_prev(:ncol,klev)) / dt`

Otherwise this routine is needlessly converting temperature to thermal energy, and then immediately converting that thermal energy back into temperature.

Reply:
Agreed, this looks like a leftover from the current CAM `ptend` vs. `tend` split where `ptend` calculated a heating rate while `tend` stored a temperature tendency. Of course, this gets me thinking that maybe the standard_name for `ttend_t` is not quite right since it may not be a "total" temperature tendency all by itself.
I see no reason to avoid this change. I do not think that kessler_update will called from CAM.

Reply:
I decided to not change this.  At this point, I'm trying to maintain the "original code" as much as possible for traceability.  Feel free to point out that you disagree that this is not the way to approach the transition.

Reply:
@cacraigucar, @nusbaume: Please do not "Resolve conversation" unless you are the conversation starter. This makes sure that the issue is resolved to the satisfaction of the reviewer before hiding it from view.
Responding to your point, while I agree in principle that we should not make changes for the fun of it, this conversion process represents an unprecedented opportunity to clean up our code. Imagine a graduate student, years from now, getting confused by why the code is this way, or worse, using it as an example of 'good parameterization coding style' and propagating it. After all, the simpler physics code is what is taught in school.
On the third hand, this very particular case is likely to go away when we no longer need to manually back out tendencies.
On the fourth hand, when that happens, it would be nice to obtain the same answer as the current manual process.

Reply:
So which hand wins?  

Also, the reason I "resolve conversation" is that on big commits (like the cam_snapshot), I totally get lost with what I had completed and what was still outstanding.  I got into the habit of "resolving conversations" when I had addressed it.  BTW, I'm not sure if all reviewers will go back and "resolve conversations".  I guess I can add my own special keyword if you are really against my workflow.

Reply:
I think I am with 'The Fourth Hand' (sounds like some sort of thriller or horror movie).

We should discuss the workflow, however, just because you (or any of us) have 'addressed' an issue does not mean is has been 'completed'. Once you click 'Resolve conversation', it then becomes even harder for the reviewer to understand what to focus on at re-review time.

How about adding a entry to the conversation specifying what you did to address the issue? Then, any conversation where you (currently) have the last word is something you do not have to worry about until the re-reviews come in.

Reply:
OK - I'll just ignore the resolve conversation button.  

Drat - you chose the fourth hand and I'm not sure what you mean by it.   If you are requesting a change, let me know.  I assume it would be to implement Jesse's suggestion.  If you are not requesting a change, then I believe I'm done.

Reply:
See, I told you "The Fourth Hand" was a horror movie waiting to be made ;)
In this case, I think it means replacing lines 76 - 78 with:
```
new_temp(:ncol) = theta(:ncol,klev) * exner(:ncol,klev)
ttend_t(:ncol,klev) = ttend_t(:ncol,klev) + ((new_temp(:ncol) - temp_prev(:ncol,klev)) / dt)
```
where `new_temp` (or some similar name) replaces `ptend_s` as a local variable.

Reply:
I agree that it would be good to get rid of the unnecessary `cpair` calculations.  I am even fine with keeping the original code in, but commented out (which I realize violates all of our coding standards), if there are concerns about traceability, simply because as currently written we will be wasting cycles doing division for no reason.

Of course, if there is a concern that this modification will cause meaningful answer changes (which it shouldn't, but you never know) then I am fine to leave it as-is for now.

Reply:
Put in change as described by @gold2718.  For the 9 timestep case it did not change answers.  This will need to be revisited if answers don't agree on longer runs.

Reply:
Resolved

Reply:
Thanks!  Once these changes (and the `phis` changes below) have been added to this PR I'll go ahead and approve it.